crates/jws/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+use base64::{engine::general_purpose, Engine as _};
+use crypto::{ed25519::Ed25199, secp256k1::Secp256k1, CryptoError, CurveOperations};
+use dids::{
+    bearer::{BearerDid, BearerDidError},
+    document::{DocumentError, KeyIdFragment, KeySelector},
+    resolver::{ResolutionError, Resolver},
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(thiserror::Error, Debug, Clone, PartialEq)]
+pub enum JwsError {
+    #[error(transparent)]
+    BearerDidError(#[from] BearerDidError),
+    #[error("serialization error {0}")]
+    SerializationError(String),
+    #[error("deserialization error {0}")]
+    DeserializationError(String),
+    #[error("decoding error {0}")]
+    DecodingError(String),
+    #[error("incorrect number of parts 3 expected {0}")]
+    IncorrectPartsLength(String),
+    #[error(transparent)]
+    DocumentError(#[from] DocumentError),
+    #[error(transparent)]
+    ResolutionError(#[from] ResolutionError),
+    #[error("algorithm not found {0}")]
+    AlgorithmNotFound(String),
+    #[error(transparent)]
+    CryptoError(#[from] CryptoError),
+}
+
+pub fn splice_parts(compact_jws: &str) -> Result<Vec<String>, JwsError> {
+    let parts: Vec<String> = compact_jws.split('.').map(|x| x.to_string()).collect();
+    if parts.len() != 3 {
+        return Err(JwsError::IncorrectPartsLength(compact_jws.to_string()));
+    }
+    Ok(parts)
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct JwsHeader {
+    pub alg: String,
+    pub kid: String,
+    pub typ: String,
+}
+
+impl JwsHeader {
+    pub fn new(alg: String, kid: String, typ: String) -> Self {
+        Self { alg, kid, typ }
+    }
+
+    pub fn new_from_encoded(encoded_jws_header: &str) -> Result<Self, JwsError> {
+        let decoded_bytes = general_purpose::URL_SAFE_NO_PAD
+            .decode(encoded_jws_header)
+            .map_err(|e| JwsError::DecodingError(e.to_string()))?;
+        let jws_header = serde_json::from_slice(&decoded_bytes)
+            .map_err(|e| JwsError::DeserializationError(e.to_string()))?;
+        Ok(jws_header)
+    }
+
+    pub fn from_bearer_did(
+        bearer_did: &BearerDid,
+        key_selector: &KeySelector,
+        typ: &str,
+    ) -> Result<Self, JwsError> {
+        let verification_method = bearer_did.document.get_verification_method(key_selector)?;
+        let kid = verification_method.id;
+        let alg = match verification_method.public_key_jwk.crv.as_str() {
+            "secp256k1" => "ES256K".to_string(),
```

Comment:
n00b rust question, but is the to_string needed?

Reply:
`"this is a str which is an array of UTF-8 bytes"` and `"this is a String type which is managed on the heap".to_string()`

basically it's a distinction between `str` and `String` types -- the former is a static memory (not sure if on the stack or in the static location in memory, but either way) and the latter is on the heap

there's a lot to even this question, just rust thangs

here's a good SO answer on this https://stackoverflow.com/a/24159933